### PR TITLE
Chiara request allergies

### PR DESCRIPTION
**Pull request by akkuebler**
trouble shooting for allergie LIST/SET in profile --> so far only allergies update, cuisine is yet to come

### CODE REVIEW

File: src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java [hunk 1]
```diff
@@ -4,6 +4,9 @@
 
 import javax.persistence.*;
 import java.io.Serializable;
+import java.util.Set;
+import java.util.HashSet;
+
```
Comment (akkuebler) [inline src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java:9]: pls ignore all those empty rows - gonna clean up afterwards

File: src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java [hunk 2]
```diff
@@ -94,37 +99,48 @@ public void setStatus(UserStatus status) {
...
-    public String getFavoriteCuisine() {
-        return favoriteCuisine;
+    public void addAllergy(String allergy) {
+        if (allergies == null) {
+            allergies = new HashSet<>();
+        }
+        allergies.add(allergy);
     }
 
+    public void removeAllergies() {
```
Comment (wlany) [inline src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java:117]: would be nicer if the naming was more consistent, so getAllergies, setAllergies, addAllergy, removeAllergy, also I'm not sure the setAllergies is necessary, I would have replaced it with the add and remove

File: src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java [hunk 3]
```diff
@@ -38,8 +41,11 @@ public class User implements Serializable {
     @Column(nullable = false)
     private UserStatus status;
 
-    @Column(nullable = true)
-    private String allergies;
+  @Column(nullable = true)
```
Comment (wlany) [inline src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java:44]: is the @Column necessary, or does it not conflict with the @ElementCollection? wondering especially whether the guestIds in Group need this annotation as well ^^'

File: src/main/java/ch/uzh/ifi/hase/soprafs23/rest/dto/UserGetDTO.java [hunk 4]
```diff
@@ -1,13 +1,20 @@
...
+
+
 
 public class UserGetDTO {
 
   private Long id;
   private String username;
   private UserStatus status;
-  private String allergies;
+  private Set<String> allergiesSet;
```
Comment (wlany) [inline src/main/java/ch/uzh/ifi/hase/soprafs23/rest/dto/UserGetDTO.java:17]: I'm unsure about the renaming... why not leave it allergies, as in the entity?

File: src/main/java/ch/uzh/ifi/hase/soprafs23/service/UserService.java [hunk 5]
```diff
@@ -140,9 +142,11 @@ public void updateUser(Long id, UserPutDTO userPutDTO,String currentPassword) {
         }
 
         if(newAllergies != null) {
-            user.setAllergies(newAllergies);
+            Set<String> allergiesSet = new HashSet<>(newAllergies); //convert here from List to Set to not make it in UserPutDTO //TODO: not necessary anymore?
```
Comment (wlany) [inline src/main/java/ch/uzh/ifi/hase/soprafs23/service/UserService.java:145]: as the TODO says, I don't think this is necessary as newAllergies is already a set not a list

File: src/test/java/ch/uzh/ifi/hase/soprafs23/controller/UserControllerTest.java [hunk 6]
```diff
@@ -264,6 +271,13 @@ public void postUsersUserIdLogout_unauthorized() throws Exception {
 
     @Test
     public void testGetUserByIdReturns200() throws Exception {
+        User user = new User();
+        user.setId(1L);
+        user.setUsername("firstUsername");
+        user.setPassword("firstPassword");
+        user.setToken("firstToken");
+        user.setStatus(UserStatus.ONLINE);
+        user.setAllergiesSet(new HashSet<>(Arrays.asList("garlic", "nuts")));
```
Comment (wlany) [inline src/test/java/ch/uzh/ifi/hase/soprafs23/controller/UserControllerTest.java:280]: aren't these lines in the @BeforeEach? I think only the last one setting the allergies is necessary for this test, and maybe putting that one in the @BeforeEach as well would be better

File: src/test/java/ch/uzh/ifi/hase/soprafs23/service/UserServiceTest.java [hunk 7]
```diff
@@ -198,7 +196,7 @@ public void testUpdateUserWithExistingUsername() {
         userPutDTO.setUsername(newUsername);
 
         when(userRepository.findById(userId)).thenReturn(Optional.of(existingUser));
-        when(userRepository.findByUsername(newUsername)).thenReturn(new User());
+        when(userRepository.findByUsername(newUsername)).thenReturn(new User()); //this was before: new User()
```
Comment (wlany) [inline src/test/java/ch/uzh/ifi/hase/soprafs23/service/UserServiceTest.java:199]: seeing the changes side by side: you only added the comment here ^^'

Comment (akkuebler) [reply src/test/java/ch/uzh/ifi/hase/soprafs23/service/UserServiceTest.java:199]: cleaned it up

File: src/test/java/ch/uzh/ifi/hase/soprafs23/controller/UserControllerTest.java (hunk 6 above)
Comment (akkuebler) [reply src/test/java/ch/uzh/ifi/hase/soprafs23/controller/UserControllerTest.java:280]: jep, leftover from writing the tests new. changed/shorten it

File: src/main/java/ch/uzh/ifi/hase/soprafs23/service/UserService.java (hunk 5 above)
Comment (akkuebler) [reply src/main/java/ch/uzh/ifi/hase/soprafs23/service/UserService.java:145]: jep, cleaned it up

File: src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java (hunk 2 above)
Comment (akkuebler) [reply src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java:117]: I like it that we have clearly stated that its a set. But I alligned it a bit to have now: getAllergiesSet, setAllergiesSet, addAllergy, removeAllergy

File: src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java (hunk 3 above)
Comment (akkuebler) [reply src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java:44]: you mean bc it is displayed in greyish? I guess it is necessary and would leave it for now: 
grafik

File: src/main/java/ch/uzh/ifi/hase/soprafs23/rest/dto/UserGetDTO.java (hunk 4 above)
Comment (akkuebler) [reply src/main/java/ch/uzh/ifi/hase/soprafs23/rest/dto/UserGetDTO.java:17]: I like that its highlighting being a set. So we have allergy and AllergiesSet. Helps me understanding the code, especially while having this here: 
grafik

File: src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java [hunk 8]
```diff
@@ -39,21 +42,19 @@ public class User implements Serializable {
     private UserStatus status;
 
     @Column(nullable = true)
-    private String allergies;
+    @ElementCollection
+    private Set<String> allergies = new HashSet<>(); // important to initialize as an empty set
```
Comment (wlany) [inline src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java:46]: I think it's a good point with the making clear that it is a set, maybe then renaming the field to allergiesSet here? following the naming scheme of the DTOs as well

Comment (akkuebler) [reply src/main/java/ch/uzh/ifi/hase/soprafs23/entity/User.java:46]: alright! adjusted

**Pull request merged**